libtenzir/builtins/formats/syslog.cpp

```diff
@@ -18,88 +18,160 @@ namespace tenzir::plugins::syslog {
 
 namespace {
 
+template <typename Message>
+struct syslog_row {
+  syslog_row(Message msg, size_t line_no)
+    : parsed(std::move(msg)), starting_line_no(line_no) {
+  }
+
+  void emit_diag(std::string_view parser_name, diagnostic_handler& diag) const {
+    if (line_count == 1) {
+      diagnostic::error("syslog parser ({}) failed", parser_name)
+        .note("input line number {}", starting_line_no)
+        .emit(diag);
+      return;
+    }
+    diagnostic::error("syslog parser ({}) failed", parser_name)
+      .note("input lines number {} to {}", starting_line_no,
+            starting_line_no + line_count - 1)
+      .emit(diag);
+  }
+
+  Message parsed;
+  size_t starting_line_no;
+  size_t line_count{1};
+};
+
 struct syslog_builder {
 public:
   using message_type = format::syslog::message;
+  using row_type = syslog_row<message_type>;
+
+  syslog_builder() = default;
 
-  syslog_builder() : builder_(format::syslog::make_syslog_type()) {
+  auto add_new(row_type&& row) -> void {
+    rows_.emplace_back(std::move(row));
   }
 
-  auto add(const message_type& msg) -> bool {
-    record r{
-      {"facility", msg.hdr.facility},     {"severity", msg.hdr.severity},
-      {"version", msg.hdr.version},       {"timestamp", msg.hdr.ts},
-      {"hostname", msg.hdr.hostname},     {"app_name", msg.hdr.app_name},
-      {"process_id", msg.hdr.process_id}, {"message_id", msg.hdr.msg_id},
-      {"structured_data", msg.data},      {"message", msg.msg},
-    };
-    return builder_.try_data(r).operator bool();
+  auto add_line_to_latest(std::string_view line) -> void {
+    TENZIR_ASSERT(not rows_.empty());
+    auto& latest = rows_.back();
+    if (not latest.parsed.msg) {
+      latest.parsed.msg.emplace(line);
+    } else {
+      latest.parsed.msg->push_back('\n');
+      latest.parsed.msg->append(line);
+    }
+    ++latest.line_count;
   }
 
   auto rows() -> size_t {
-    return static_cast<size_t>(builder_.length());
+    return rows_.size();
   }
 
-  auto finish() -> std::vector<table_slice> {
-    return builder_.finish_as_table_slice();
+  auto finish(diagnostic_handler& diag) -> std::optional<table_slice> {
+    series_builder builder{format::syslog::make_syslog_type()};
+    for (auto& row : rows_) {
+      auto& msg = row.parsed;
+      const record r{
+        {"facility", msg.hdr.facility},
+        {"severity", msg.hdr.severity},
+        {"version", msg.hdr.version},
+        {"timestamp", msg.hdr.ts},
+        {"hostname", std::move(msg.hdr.hostname)},
+        {"app_name", std::move(msg.hdr.app_name)},
+        {"process_id", msg.hdr.process_id},
+        {"message_id", msg.hdr.msg_id},
+        {"structured_data", std::move(msg.data)},
+        {"message", std::move(msg.msg)},
+      };
+      if (not builder.try_data(r)) {
+        row.emit_diag("RFC 5242", diag);
+        return std::nullopt;
+      }
+    }
+    rows_.clear();
+    return builder.finish_assert_one_slice();
   }
```

Comment:
The call to `finish_assert_one_slice` is dangerous because you might get more than one slice if there is a type mismatch. For example, try this:
```
<165>8 2003-10-11T22:14:15.003Z foo bar - ID47 [example@32473 iut="5"]
<165>8 2003-10-11T22:14:15.003Z foo bar - ID47 [example@32473 iut="foo"]
```